Let organisations feature policies on their homepages

edds opened the pull request:
As an organisation I want to feature the most important and high profile polices we are working on, on our homepage. This will let users of our high profile policies be able to find the content they want quicker.

---

This adds a new tab in the organisation admin to let and organisation pick their featured policies:

![screen shot 2015-06-24 at 13 38 54](https://cloud.githubusercontent.com/assets/35035/8330070/a6a4c338-1a76-11e5-8cea-004e586564c2.png)

Which are then displayed on their homepages in the order specified.

<table>
<tr>
<th>Number 10's page</th>
<th>Standard departments page</th>
</tr>
<tr>
<td><img src="https://cloud.githubusercontent.com/assets/35035/8330076/b0776604-1a76-11e5-80f9-33da8d1e41fa.png"></td><td><img src="https://cloud.githubusercontent.com/assets/35035/8330077/b2088bf6-1a76-11e5-968d-6d1df16bccef.png"></td>
</tr>
</table>

There is a data migration which will ensure that any organisation which currently has policies showing on their homepages will continue to have the same policies listed. They can then be manually edited in the admin if they want a different collection of policies.

elliotcm commented:
So the new tab is "Features"?

elliotcm commented on app/controllers/admin/featured_policies_controller.rb:
```diff
@@ -0,0 +1,41 @@
+  end
+
+  def create
+    @featured_policies = @organisation.featured_policies.build(policy_content_id: params[:policy_content_id])
```
`featured_policy`, right?

elliotcm replied:
Also you don't need to use an instance variable here.

edds replied:
Yer, that is what it says ;)

elliotcm replied:
It says "featured_policies" but you're only building one.

edds commented:
No the new tab is ['Featured policies'](https://github.com/alphagov/whitehall/pull/2243/files#diff-595bc510088e3a7f1da5efc25e7cf7fdR38) it is hidden under the more link.

elliotcm commented on app/controllers/admin/featured_policies_controller.rb:
```diff
@@ -0,0 +1,41 @@
+    if @featured_policies.save
+      redirect_to admin_organisation_featured_policies_path(@organisation), notice: 'Policies featured'
+    else
+      render :new
```
You don't have a view for this action.

elliotcm commented on app/controllers/admin/featured_policies_controller.rb:
```diff
@@ -0,0 +1,41 @@
+  end
+
+  def destroy
+    @feature = @organisation.featured_policies.find(params[:id])
```
Don't need an instance variable, and we should make the name consistent with the `create` action.

edds replied:
updated.

elliotcm commented on app/controllers/admin/featured_policies_controller.rb:
```diff
@@ -0,0 +1,41 @@
+
+  def destroy
+    @feature = @organisation.featured_policies.find(params[:id])
+    if @feature.destroy!
```
This raises an exception.  You want `#destroy`.

edds replied:
Updated

elliotcm commented on app/controllers/admin/featured_policies_controller.rb:
```diff
@@ -0,0 +1,41 @@
+  def reorder
+    params[:ordering].each_pair do |featured_policy_id, order|
+      FeaturedPolicy.find(featured_policy_id).update_attributes(
+        ordering: order
```
This will fit on a single line.  If you've broken it out to make possible future diffs clearer, end this with a comma.

edds replied:
Moved onto a single line.

elliotcm commented on app/views/admin/featured_policies/_feature_list.html.erb:
```diff
@@ -0,0 +1,34 @@
+              <%= link_to(admin_organisation_featured_policy_path(organisation, featured_policy),
+                    data: { confirm: "Unfeature '#{featured_policy.title}'?" },
+                    method: :delete,
+                    class: "btn btn-danger") do %>Unfeature <span class="visuallyhidden"><%= featured_policy.title %></span><% end %>
```
Some newlines would make this more readable.

elliotcm commented on app/views/admin/featured_policies/_policy_list.html.erb:
```diff
@@ -0,0 +1,21 @@
+        <td>
+          <%= button_to(admin_organisation_featured_policies_path(policy_content_id: policy.content_id),
+                method: :post,
+                class: 'btn') do %>Feature <span class="visuallyhidden"><%= policy.title %></span><% end %>
```
Some newlines would make this more readable.

elliotcm commented on db/data_migration/20150624075140_populate_featured_policies.rb:
```diff
@@ -0,0 +1,41 @@
+    orgs = policy["links"]["organisations"]
+    orgs.each { |org| memo[org["content_id"]] = 1 } if orgs
+    memo
+  }.keys
```
I don't really understand what this is doing.

edds replied:
This was a way of getting a uniq collection of orgs. It turns out ruby can just do that on an array for you though. Seems I have been in JS land for to long. Updated, hopefully its clearer.

elliotcm commented on db/data_migration/20150624075140_populate_featured_policies.rb:
```diff
@@ -0,0 +1,41 @@
+def policy_content_id(policies, policy_link)
+  policy = policies.select do |policy|
+    policy["base_path"] == policy_link
+  end.first
```
Chaining do/end is against [the styleguide](https://github.com/alphagov/styleguides/blob/master/ruby.md#syntax), use braces.

evilstreak replied:
Or better still, use [`Enumerable#find`](http://ruby-doc.org/core-2.1.6/Enumerable.html#method-i-find)

elliotcm replied:
Yeah that.

elliotcm commented on features/organisations.feature:
```diff
@@ -213,3 +213,27 @@ Scenario: Citizen views a closed organisation
+  When I stop featuring the polices "Sing aloud" for "Department of Fun"
+  Then there should be no featured policies on the home page of "Department of Fun"
+
+Scenario: Featuring policies on an organisation
```
This scenario title is the same as the other one.

elliotcm commented on test/factories/featured_policies.rb:
```diff
@@ -0,0 +1,6 @@
+FactoryGirl.define do
+  factory :featured_policy do
+    organisation
+    policy_content_id ''
```
Factories should create valid objects, but this isn't properly valid with a blank content ID.

elliotcm commented:
This is generally pretty good, but comments aside I've a few worries.
1. This is removing policies from all org homepages except number 10, right?  I don't think we've discussed that with them.
2. This is changing the number of policies on homepages from 3 to 5.
3. There are a number of branches in the code which were broken but the tests still passed, namely the "what happens if a save fails" branches, which suggests they're not covered by any tests.
4. Minor point, but some newlines around nested blocks (like `if`s) in methods would make things easier to follow.

elliotcm commented:
Also, we've been pushing towards a style of writing features which doesn't jump straight to the capybara instructions in the step definitions.  @evilstreak is this still a thing we care about?

evilstreak commented:
If we're going to go to the trouble of using Cucumber (rather than another test framework for integration tests, like Rspec feature specs) we ought to get the right levels of abstraction in each of the three layers.

So yes, I think we still care about that!

edds commented:
> 1. This is removing policies from all org homepages except number 10, right? I don't think we've discussed that with them.

Once this has run there should be no public difference on the site. The data migration ensures that all orgs with policies will continue to have the same policies listed.

> 2. This is changing the number of policies on homepages from 3 to 5.

This is as requested in the feature. We removed the descriptions and that now makes 3 feel like too few, so we are bumping it up to 5.

> 3. There are a number of branches in the code which were broken but the tests still passed, namely the "what happens if a save fails" branches, which suggests they're not covered by any tests.

I believe that is the only branch isn't it. Not sure what would cause the save to fail in order to write a test. I can remove that branch though.

> Also, we've been pushing towards a style of writing features which doesn't jump straight to the capybara instructions in the step definitions.

If this is so then it should be documented. Documenting how we want tests in this repo at all would be really useful, there are so many test all over the place with all sorts of different styles. I was copying the style used by the other organisation features for this.

elliotcm commented:
> Once this has run there should be no public difference on the site. The data migration ensures that all orgs with policies will continue to have the same policies listed.

I've traced through the partials and you're right, this is fine.

> This is as requested in the feature. We removed the descriptions and that now makes 3 feel like too few, so we are bumping it up to 5.

:+1: 

> I believe that is the only branch isn't it. Not sure what would cause the save to fail in order to write a test. I can remove that branch though.

There's one when trying to create and one when trying to delete.  Failing validation is a good way to test the creation, less sure about deletion as we might not have any callbacks that would stop it short of a database failure.

> If this is so then it should be documented. Documenting how we want tests in this repo at all would be really useful, there are so many test all over the place with all sorts of different styles. I was copying the style used by the other organisation features for this.

This is true, it's not currently documented.  We should sort that.  In the meantime you can use something like the policy tagging feature as a guide:
- https://github.com/alphagov/whitehall/blob/master/features/admin-policy-tagging.feature
- https://github.com/alphagov/whitehall/blob/master/features/step_definitions/admin_policy_tagging_steps.rb
- https://github.com/alphagov/whitehall/blob/master/features/support/admin_policy_tagging_helper.rb

elliotcm commented on db/schema.rb:
```diff
@@ -525,6 +525,12 @@
 
+  create_table "featured_policies", force: :cascade do |t|
+    t.string  "policy_content_id", limit: 255
+    t.integer "ordering",          limit: 4
```
Something else just occurred to me; we wound up in a bunch of trouble with attachment orderings a while back which was exacerbated by allowing the ordering field to be `null`.  Since then we've had constraints on those fields to [require an ordering](https://github.com/alphagov/whitehall/blob/master/db/schema.rb#L66).

edds replied:
It shouldn't ever be able to get in a null state due to the [before save](https://github.com/alphagov/whitehall/commit/06978b429f22c8ff78db3021047e69c22d9c4cc3#diff-1f693d12e26b3550ceb2198d1eaf824bR6) on the object adding a value, but have added the constraint.

edds commented:
I have removed the code paths which were untested. I don't think the code will have ever got in those states and now it should exception at us instead.

I have updated all the cucumbers, hopefully they are clearer now.

elliotcm commented:
That's fine as long as we don't have any validations, otherwise the user will get a 500 instead of feedback on the form.

elliotcm commented:
:+1: when green
